### PR TITLE
[new release] decoders, decoders-yojson, decoders-sexplib, decoders-msgpck, decoders-jsonm, decoders-ezjsonm, decoders-cbor and decoders-bencode (0.7.0)

### PR DESCRIPTION
**Pull request by mattjbray**
Elm-inspired decoders for Ocaml

- Project page: <a href="https://github.com/mattjbray/ocaml-decoders">https://github.com/mattjbray/ocaml-decoders</a>
- Documentation: <a href="https://mattjbray.github.io/ocaml-decoders/">https://mattjbray.github.io/ocaml-decoders/</a>

##### CHANGES:

* Add `Decode.field_opt_or` (mattjbray/ocaml-decoders#43, @c-cube)
* Add `Decode.pick` (mattjbray/ocaml-decoders#43, @c-cube)
* Add `Decode.decode_sub` (mattjbray/ocaml-decoders#45, @c-cube)

### `bs-decoders`

* Improve `int` decoder (@actionshrimp)

### `decoders-yojson`

* Move away from deprecated `Yojson.json` to `Yojson.t` (mattjbray/ocaml-decoders#37, @idkjs)

### CODE REVIEW

File: packages/decoders-bencode/decoders-bencode.0.7.0/opam [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+maintainer: ["Matt Bray <mattjbray@gmail.com>"]
+authors: ["Simon Cruanes <simon@imandra.ai>"]
+license: "ISC"
+homepage: "https://github.com/mattjbray/ocaml-decoders"
+doc: "https://mattjbray.github.io/ocaml-decoders/"
+bug-reports: "https://github.com/mattjbray/ocaml-decoders/issues"
+depends: [
+  "ocaml" {>= "4.03.0"}
+  "dune" {>= "2.0"}
+  "decoders"
```
[the remark this answers is not in the document]
Comment (mattjbray) [reply packages/decoders-bencode/decoders-bencode.0.7.0/opam]: Thanks! I've added and done another `dune-release`

File: packages/decoders-bencode/decoders-bencode.0.7.0/opam [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+authors: ["Simon Cruanes <simon@imandra.ai>"]
+license: "ISC"
+homepage: "https://github.com/mattjbray/ocaml-decoders"
+doc: "https://mattjbray.github.io/ocaml-decoders/"
+bug-reports: "https://github.com/mattjbray/ocaml-decoders/issues"
+depends: [
+  "ocaml" {>= "4.03.0"}
+  "dune" {>= "2.0"}
+  "decoders" {= version}
+  "bencode"
```
Comment (mseri) [inline packages/decoders-bencode/decoders-bencode.0.7.0/opam:15]:
```suggestion
  "bencode" {>= "0.2"}
```
For
```
# (cd _build/default && /home/opam/.opam/4.03/bin/ocamlopt.opt -w -40 -g -I src-bencode/.decoders_bencode.objs/byte -I src-bencode/.decoders_bencode.objs/native -I /home/opam/.opam/4.03/lib/bencode -I /home/opam/.opam/4.03/lib/decoders -intf-suffix .ml -no-alias-deps -open Decoders_bencode -o src-bencode/.decoders_bencode.objs/native/decoders_bencode__Decode.cmx -c -impl src-bencode/decode.ml)
# File "src-bencode/decode.ml", line 28, characters 27-30:
# Error: This expression has type int but an expression was expected of type
#          int64
```

File: packages/decoders-msgpck/decoders-msgpck.0.7.0/opam [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  ["dune" "subst"] {pinned}
+  [
+    "dune"
+    "build"
+    "-p"
+    name
+    "-j"
+    jobs
+    "@install"
+    "@runtest" {with-test}
```
Comment (mseri) [inline packages/decoders-msgpck/decoders-msgpck.0.7.0/opam:33]:
```suggestion
    "@runtest" {with-test & ocaml:version >= "4.08"}
```
For
```
runtest".

#=== ERROR while compiling decoders-msgpck.0.7.0 ==============================#
# context              2.1.2 | linux/x86_64 | ocaml-base-compiler.4.06.1 | pinned(https://github.com/mattjbray/ocaml-decoders/releases/download/v0.7.0/decoders-0.7.0.tbz)
# path                 ~/.opam/4.06/.opam-switch/build/decoders-msgpck.0.7.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p decoders-msgpck -j 31 @install @runtest
# exit-code            1
# env-file             ~/.opam/log/decoders-msgpck-1042-2676d4.env
# output-file          ~/.opam/log/decoders-msgpck-1042-2676d4.out
### output ###
# (cd _build/default && /home/opam/.opam/4.06/bin/ocamlc.opt -w -40 -g -bin-annot -I test-msgpck/.main.eobjs/byte -I /home/opam/.opam/4.06/lib/containers -I /home/opam/.opam/4.06/lib/containers/monomorphic -I /home/opam/.opam/4.06/lib/decoders -I /home/opam/.opam/4.06/lib/either -I /home/opam/.opam/4.06/lib/msgpck -I /home/opam/.opam/4.06/lib/oUnit -I /home/opam/.opam/4.06/lib/ocplib-endian -I /home/opam/.opam/4.06/lib/ounit2 -I /home/opam/.opam/4.06/lib/ounit2/advanced -I /home/opam/.opam/4.06/lib/seq -I /home/opam/.opam/4.06/lib/stdlib-shims -I src-msgpck/.decoders_msgpck.objs/byte -no-alias-deps -o test-msgpck/.main.eobjs/byte/dune__exe__Main.cmo -c -impl test-msgpck/main.ml)
# File "test-msgpck/main.ml", line 73, characters 39-40:
# Error: Syntax error: operator expected.
```

File: packages/decoders-bencode/decoders-bencode.0.7.0/opam [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+authors: ["Simon Cruanes <simon@imandra.ai>"]
+license: "ISC"
+homepage: "https://github.com/mattjbray/ocaml-decoders"
+doc: "https://mattjbray.github.io/ocaml-decoders/"
+bug-reports: "https://github.com/mattjbray/ocaml-decoders/issues"
+depends: [
+  "ocaml" {>= "4.03.0"}
+  "dune" {>= "2.0"}
+  "decoders" {= version}
+  "bencode" {>= "0.2"}
```
Comment (mseri) [inline packages/decoders-bencode/decoders-bencode.0.7.0/opam:15]:
```suggestion
  "bencode" {>= "2.0"}
```
The breaking change for int64 was introduced in version 2

File: packages/decoders-bencode/decoders-bencode.0.7.0/opam [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  "ocaml" {>= "4.03.0"}
+  "dune" {>= "2.0"}
+  "decoders" {= version}
+  "bencode" {>= "2.0"}
+  "odoc" {with-doc}
+  "containers" {with-test & >= "0.16"}
+  "ounit" {with-test}
+]
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/decoders-bencode/decoders-bencode.0.7.0/opam:21]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/decoders-cbor/decoders-cbor.0.7.0/opam [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+depends: [
+  "ocaml" {>= "4.03.0"}
+  "dune" {>= "2.0"}
+  "decoders" {= version}
+  "cbor"
+  "odoc" {with-doc}
+  "containers" {with-test & >= "0.16"}
+  "ounit" {with-test}
+]
+build: [
```
Comment (mseri) [inline packages/decoders-cbor/decoders-cbor.0.7.0/opam:20]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/decoders-ezjsonm/decoders-ezjsonm.0.7.0/opam [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  "ocaml" {>= "4.03.0"}
+  "dune" {>= "2.0"}
+  "decoders" {= version}
+  "ezjsonm" {>= "0.4.0"}
+  "odoc" {with-doc}
+  "containers" {with-test & >= "0.16"}
+  "ounit" {with-test}
+]
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/decoders-ezjsonm/decoders-ezjsonm.0.7.0/opam:21]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/decoders-jsonm/decoders-jsonm.0.7.0/opam [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  "ocaml" {>= "4.03.0"}
+  "dune" {>= "2.0"}
+  "decoders" {= version}
+  "jsonm"
+  "odoc" {with-doc}
+  "containers" {with-test & >= "0.16"}
+  "ounit" {with-test}
+]
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/decoders-jsonm/decoders-jsonm.0.7.0/opam:21]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/decoders-msgpck/decoders-msgpck.0.7.0/opam [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  "dune" {>= "2.0"}
+  "decoders" {= version}
+  "msgpck" {>= "1.3"}
+  "ocplib-endian" {>= "0.6"}
+  "odoc" {with-doc}
+  "containers" {with-test & >= "0.16"}
+  "ounit" {with-test}
+]
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/decoders-msgpck/decoders-msgpck.0.7.0/opam:24]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/decoders-sexplib/decoders-sexplib.0.7.0/opam [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  "dune" {>= "2.0"}
+  "decoders" {= version}
+  "sexplib0"
+  "sexplib"
+  "odoc" {with-doc}
+  "containers" {with-test & >= "0.16"}
+  "ounit" {with-test}
+]
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/decoders-sexplib/decoders-sexplib.0.7.0/opam:22]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/decoders-yojson/decoders-yojson.0.7.0/opam [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+  "ocaml" {>= "4.03.0"}
+  "dune" {>= "2.0"}
+  "decoders" {= version}
+  "yojson" {>= "1.6.0"}
+  "odoc" {with-doc}
+  "containers" {with-test & >= "0.16"}
+  "ounit" {with-test}
+]
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/decoders-yojson/decoders-yojson.0.7.0/opam:21]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/decoders/decoders.0.7.0/opam [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+doc: "https://mattjbray.github.io/ocaml-decoders/"
+bug-reports: "https://github.com/mattjbray/ocaml-decoders/issues"
+depends: [
+  "ocaml" {>= "4.03.0"}
+  "dune" {>= "2.0"}
+  "odoc" {with-doc}
+  "containers" {with-test & >= "0.16"}
+]
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/decoders/decoders.0.7.0/opam:18]:
```suggestion
  ["dune" "subst"] {dev}
```

**Review by mseri — COMMENTED** [review]
It is good now. The only remaining issue is a bug in dune (fixed in dune Lang >= 2.7) but we can simply fix the packages oN the opam repository. If you want this can be merged once those fixes are committed to this PR, you don’t need to retag and rerelease since the opam files here are the ones used by opam

**Comment by mattjbray** [comment]
```
#=== ERROR while fetching sources for decoders-ezjsonm.0.5.0 ==================#
OpamSolution.Fetch_fail("Sys_error(\"rm -rf '/tmp/opam-990-465eb8': Cannot allocate memory\")")
```
🤔

**Comment by mseri** [comment]
CI problem, don't worry about that. Can I merge or you prefer to re-release?

**Comment by mattjbray** [comment]
Yes, please merge. I've made the same changes upstream so hopefully things will be smoother on the next release 🤞 . Thanks for your help on this.

**Comment by mseri** [comment]
Thanks for your patience

**Pull request merged**